client-core/src/service/wallet_service.rs

```diff
@@ -177,52 +176,159 @@ impl Wallet {
-    /// find staking key
-    pub fn find_staking_key(&self, redeem_address: &RedeemAddress) -> Option<&PublicKey> {
-        self.staking_keys
-            .iter()
-            .find(|staking_key| &RedeemAddress::from(*staking_key) == redeem_address)
+fn read_pubkey<S: SecureStorage>(storage: &S, keyspace: &str, key: &str) -> Result<PublicKey> {
+    let value = storage.get(keyspace, key)?;
+    if let Some(raw_value) = value {
+        let pubkey = PublicKey::deserialize_from(&raw_value)?;
+        Ok(pubkey)
+    } else {
+        Err(Error::new(ErrorKind::InvalidInput, "read pubkey error"))
     }
+}
 
-    /// find private key
-    pub fn find_private_key(&self, public_key: &PublicKey) -> Option<&PrivateKey> {
-        self.key_pairs.get(public_key)
+fn write_pubkey<S: SecureStorage>(
+    storage: &S,
+    keyspace: &str,
+    key: &str,
+    value: &PublicKey,
+) -> Result<()> {
+    storage.set(keyspace, key, value.serialize())?;
+    Ok(())
+}
+
+fn read_pubkey_enc<S: SecureStorage>(
+    storage: &S,
+    keyspace: &str,
+    key: &str,
+    enckey: &SecKey,
+) -> Result<PublicKey> {
+    let value = storage.get_secure(keyspace, key, enckey)?;
+    if let Some(raw_value) = value {
+        let pubkey = PublicKey::deserialize_from(&raw_value)?;
+        Ok(pubkey)
+    } else {
+        Err(Error::new(ErrorKind::InvalidInput, "read pubkey error"))
     }
+}
+
+fn write_pubkey_enc<S: SecureStorage>(
+    storage: &S,
+    keyspace: &str,
+    key: &str,
+    value: &PublicKey,
+    enckey: &SecKey,
+) -> Result<()> {
+    storage.set_secure(keyspace, key, value.serialize(), enckey)?;
+    Ok(())
+}
 
-    /// add key pair
-    pub fn add_key_pair(&mut self, public_key: &PublicKey, private_key: &PrivateKey) {
-        self.key_pairs
-            .insert(public_key.clone(), private_key.clone());
+fn read_string<S: SecureStorage>(storage: &S, keyspace: &str, key: &str) -> Result<String> {
+    let value = storage.get(keyspace, key.as_bytes())?;
+    if let Some(raw_value) = value {
+        let ret = str::from_utf8(&raw_value).chain(|| {
+            (
+                ErrorKind::InvalidInput,
+                "Unable to read string in wallet_service",
+            )
+        })?;
+        Ok(ret.to_string())
+    } else {
+        Err(Error::new(ErrorKind::InvalidInput, "read string error"))
     }
+}
 
-    /// find root hash
-    pub fn find_root_hash(&self, address: &ExtendedAddr) -> Option<&H256> {
-        match address {
-            ExtendedAddr::OrTree(ref root_hash) => {
-                self.root_hashes.iter().find(|hash| hash == &root_hash)
-            }
-        }
+fn read_number<S: SecureStorage>(
+    storage: &S,
+    keyspace: &str,
+    key: &str,
+    defaut_value: Option<u64>,
+) -> Result<u64> {
+    let value = storage.get(keyspace, key.as_bytes())?;
+    if let Some(raw_value) = value {
+        let mut v: [u8; 8] = [0; 8];
+        v.copy_from_slice(&raw_value);
+        let index_value: u64 = u64::from_be_bytes(v);
```

Comment:
any special reason why big endian (note: almost everywhere else in the code, little endian is used)?

Reply:
no special reason, little endian is also good